[Doc] Build StarRocks on Ubuntu (#12744)
StarRocks/starrocks: docs/developers/build-starrocks/build_starrocks_on_ubuntu.md

### Prerequisite

```
sudo apt-get install automake binutils-dev bison byacc ccache flex libiberty-dev libtool maven zip python3 python-is-python3 -y
```

> ### Compiler

Reviewer: leave a empty line between title and content

If the ubuntu version >= 22.04, you can
```
sudo apt-get install cmake gcc g++ default-jdk -y
```

If the ubuntu version < 22.04.
Check the version of following tools and compilers

##### 1. GCC/G++
GCC/G++ version must be >= 10.3
```
gcc --version
g++ --version
```
Install GCC/G++(https://gcc.gnu.org/releases.html)

##### 2. JDK

> OpenJDK version must be >= 8

Reviewer: It should emphasize that JDK is required, not JRE, and tell people how to identify which is which.

```
java --version
```
Install OpenJdk(https://openjdk.org/install)

##### 3. CMake
cmake version must be >= 3.20.1

```
cmake --version
```
Install cmake(https://cmake.org/download)

### Improve the compile speed

> The default compiling paralleilsim equals to the **CPU Cores / 4**. If you want to improve the compile speed. You can improve the paralleilsim. Suppose you have 32 Cpu cores, want to use 24 cores to compile.

Reviewer: I think you should provide a normal build method. Otherwise, people will copy the only example `./build.sh -j 24`. Which will cause problems.

```
./build.sh -j 24
```

### FAQ
1. Failed to build `aws_cpp_sdk` in the Ubuntu 20.04.
```
Error: undefined reference to pthread_create
```

> I encountered the same error with CMake 3.16.3; you can upgrade the CMake version to at least 3.20.1

Reviewer: should change this sentence, because this is a manual, should not be A `I`

### Prerequisite
```

> sudo apt-get install automake binutils-dev bison byacc ccache cmake flex libiberty-dev libtool maven zip -y

Reviewer 2: If the versions of CMAKE and GCC are too low, they will not compile successfully. You should write the basic requirements or check in the script.

Author: cmake 3.16.3, or gcc 9.4. Larger than them?

Reviewer 2: For me, the version I compiled myself is cmake 3.20.3, and gcc is 10.3.0. Change the script and compile it successfully. The actual minimum requirements are not very clear, but I remember that the cmake version that comes with ubuntu is definitely not good.

```

### Compiler
Check the compilers(GCC, Python, JDK) are installed in the machine, if not install them
```

> sudo apt-get install gcc g++ default-jdk python3 python-is-python3 -y

Reviewer 2: Only relying on apt-get is not necessarily able to install the appropriate version of gcc, cmake or python and other tools.

Author: If download the binary failed, we can add it in the FAQ.

```

> ### Improve the compile speed
> The default compiling paralleilsim equals to the **CPU Cores / 4**. If you want to improve the compile speed. You can improve the paralleilsim. Suppose you have 32 Cpu cores, want to use 24 cores to compile.

Reviewer: ```suggestion
### Improve the compile speed

The default compiling parallelism equals to the **CPU Cores / 4**. If you want to improve the compile speed. You can improve the parallelism. Suppose you have 32 Cpu cores, and want to use 24 cores to compile.
```

```
./build.sh -j 24
```

### FAQ
1. Failed to build `aws_cpp_sdk` in the Ubuntu 20.04.
```
Error: undefined reference to pthread_create
```

> Apparently Threads::Threads is an interface library on some system. I encountered the same error with CMake 3.10.2, and applying the following patch fixed the issue.
> ```

Reviewer: ```suggestion
Apparently Threads::Threads is an interface library on some system. I encountered the same error with CMake 3.10.2, and applying the following patch fixed the issue.

```
```

> diff --git a/cmake/dependencies.cmake b/cmake/dependencies.cmake

Reviewer 2: Why not provide a build-ubuntu.sh script directly, but let the user make a patch?

Author: The problem comes from cmake's version. I think if I upgrade the cmake, the problem will dismiss.

index 30c665341f..ac8a15a2be 100644
--- a/cmake/dependencies.cmake
+++ b/cmake/dependencies.cmake
@@ -30,6 +30,17 @@ function(compute_links lib)
     # For each direct link append it and its links
     foreach(link ${links})
         if(TARGET ${link}) # Collect only target links
+            get_target_property(link_type ${link} TYPE)
+            if(link_type STREQUAL "INTERFACE_LIBRARY")
+                # A custom property LINK_LIBRARIES_ALL can't be set to INTERFACE libraries.
+                # We have to allow Threads::Threads, which might be an interface library.
+                if(link STREQUAL "Threads::Threads")
+                    continue()
+                elseif()
+                    message(STATUS "Library '${lib}' uses link '${link}'.")
+                    message(FATAL_ERROR "Algorithm doesn't work with interface libraries.")
+                endif()
+            endif()
             compute_links(${link})
             get_target_property(link_links_all ${link} LINK_LIBRARIES_ALL)
             set_property(TARGET ${lib} APPEND PROPERTY
```

> 2. Failed to build `streamvbyte`
> ```

Reviewer: ```suggestion
2. Failed to build `streamvbyte`

```
```
